src/tribol/geom/NodalNormal.hpp

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2017-2023, Lawrence Livermore National Security, LLC and
+// other Tribol Project Developers. See the top-level LICENSE file for details.
+//
+// SPDX-License-Identifier: (MIT)
+
+#ifndef SRC_GEOM_NODALNORMAL_HPP_
+#define SRC_GEOM_NODALNORMAL_HPP_
+
+#include "tribol/config.hpp"
+
+#include "tribol/mesh/MeshData.hpp"
+#include "tribol/mesh/MethodCouplingData.hpp"
+
+namespace tribol {
+
+/**
+ * @brief Virtual base class to define the interface for nodal normal calculations
+ */
+class NodalNormal {
+ public:
+  /**
+   * @brief Destructor
+   */
+  virtual ~NodalNormal() {}
+
+  /**
+   * @brief Interface for computing and storing nodal normals on a given mesh
+   *
+   * @param mesh Mesh data
+   */
+  virtual void Compute( MeshData& mesh ) = 0;
+
+  /**
+   * @brief Gets the Jacobian data for the nodal normals
+   *
+   * @pre Must call NodalNormal::Compute() first
+   *
+   * @return MethodData reference storing Jacobian data
+   */
+  MethodData& getJacobianData() { return elem_jacobians_; }
+
+  /// @overload
+  const MethodData& getJacobianData() const { return elem_jacobians_; }
+
+ private:
+  /**
+   * @brief Stores Jacobian contributions of the nodal normals
+   */
+  MethodData elem_jacobians_;
+};
+
+/**
+ * @brief Computes nodal normals as the average of connected element normals
+ */
+class ElementAvgNormal : public NodalNormal {
+ public:
+  /**
+   * @brief Computes nodal normals as the average of connected element normals
+   *
+   * @param mesh Mesh data
+   */
+  void Compute( MeshData& mesh ) override;
+};
+
+#ifdef TRIBOL_USE_ENZYME
+
+/**
+ * @brief Computes nodal normals by averaging normal evaluated at the node of all connected elements
+ */
+class VertexAvgNormal : public NodalNormal {
+ public:
+  /**
+   * @brief Constructor
+   *
+   * @param compute_deriv Computes the Jacobian if true
+   */
+  VertexAvgNormal( bool compute_deriv = true ) : compute_deriv_( compute_deriv ) {}
+
+  /**
+   * @brief Computes nodal normals by averaging normal evaluated at the node of all connected elements
+   *
+   * @param mesh Mesh data
+   */
+  void Compute( MeshData& mesh ) override;
+
+ private:
+  /**
+   * @brief Computes the Jacobian if true
+   */
+  bool compute_deriv_;
+};
+
+// free functions for enzyme
+
+/**
+ * @brief Computes the normal direction at all nodes of the element
```

Comment:
The brief and function name don't really help me understand exactly how the normal is being computed for all nodes. Is this using the edges that share a node to compute an edge averaged normal at all nodes for a given face?

Reply:
I think the confusion is because this is at the element level.  This just computes the normal at all the nodal coordinates of the element.  Moved out of header file, since it's only implemented for Enzyme.